### PR TITLE
Fixes PublishLater on PostgreSQL

### PR DESCRIPTION
**Pull request by jtkech**
Fixes https://github.com/OrchardCMS/OrchardCore/issues/7842

It was already working on my side with sqlite, but seems to not work on PostgreSQL as commented on gitter

> "PostgreSQL: 42883 Operator does not exist: timestamp without time zone = text"
> This error is thrown in OrchardCore.PublishLater/ScheduledPublishingBackgroundTask.cs
exactly at this line
> .Query<ContentItem, PublishLaterPartIndex>(index => index.ScheduledPublishUtc < _clock.UtcNow)

So, normally in the `Migrations` we would have to change the `ScheduledPublishUtc` dbType from string to `DateTime`, in an `UpdateFrom1()`, and **also update to the right type** in `Create()` that would have to **shortcut the migration step**.

But we can't alter the column to `DateTime` because `WithType()` has an int? length parameter, and it fails even if we pass a null value.

- So here in `UpdateFrom1()` i just added a new column `ScheduledPublishUtcDateTime` with the right type, this in place of altering the existing one. And in `Create()` we only create the right column, not the previous one.

Not sure it is worth to do `CreateIndex()` on an index table that will never be so large, but there was already one

- So I added a `CreateIndex()` using the new column and also including `DocumentId` as it is planned in #8245 

    This in `Create()` and also in the new migration step `UpdateFrom1()` (that is not executed on a new schema)

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndex.cs [hunk 1]
```diff
@@ -7,7 +7,7 @@ namespace OrchardCore.PublishLater.Indexes
 {
     public class PublishLaterPartIndex : MapIndex
     {
-        public DateTime? ScheduledPublishUtc { get; set; }
+        public DateTime? ScheduledPublishUtcDateTime { get; set; }
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndex.cs:10]:
```suggestion
        public DateTime? ScheduledPublishDateTimeUtc { get; set; }
```

Maybe? just to keep the consistency with Utc at the end, is often useful

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs [hunk 2]
```diff
@@ -23,16 +24,33 @@ public int Create()
...
+            return 2;
+        }
+
+        // This code can be removed in a later version.
+        public int UpdateFrom1()
+        {
+            SchemaBuilder.AlterIndexTable<PublishLaterPartIndex>(table => table
+                .AddColumn<DateTime>(nameof(PublishLaterPartIndex.ScheduledPublishUtcDateTime))
+            );
+
```
Comment (deanmarcussen) [inline src/OrchardCore.Modules/OrchardCore.PublishLater/Migrations.cs:46]: No drop of the useless column?

**Review by deanmarcussen — COMMENTED** [review]
I can't see any other way either @jtkech but it will be breaking for any existing published later.

They will just need to be republished to catch the new correct value.

I assume it must have worked previously on SQL Server ok as well, or @Piedone would have noticed.

**Comment by Piedone** [comment]
I don't know how I managed to mess up the data type, apologies. Strange that it works with SQLite.

File: src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndex.cs (hunk 1 above)
Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.PublishLater/Indexes/PublishLaterPartIndex.cs:10]: I would have gone with ScheduledPublishUtc just like we do with CreatedUtc. Not sure there is a need for repetition of the dbtype in the name.

**Comment by jtkech** [comment]
@deanmarcussen 

> I can't see any other way either @jtkech but it will be breaking for any existing published later.
> They will just need to be republished to catch the new correct value.

I tried the migration step with sqlite (i will do with sql server), there is no problem with the new data type, the only concern is for items that are already enlisted to be published later, yes they need to be "republished later" again.

> I assume it must have worked previously on SQL Server ok as well, or @Piedone would have noticed.

I will try it for SQL server

> Suggested change ScheduledPublishUtcDateTime => ScheduledPublishDateTimeUtc

Okay, I just used `UtcDateTime` as it is the same name used for the `DateTimeOffset.UtcDateTime` property ;)

> No drop of the useless column?

As i remember you had some problem with some migration steps, and **you commented that it is not safe to remove an index, is it still true?** if so, we can't remove this useless column that is still included in an existing index

So, is it still a problem to remove an index in  migration step?

If not included in an index, is it safe to remove a column in a migration step?

@Skrypt 

> I would have gone with ScheduledPublishUtc just like we do with CreatedUtc. Not sure there is a need for repetition of the dbtype in the name.

This is exactly the name of the previous column that for now i didn't removed for the above reasons

@Piedone 

> I don't know how I managed to mess up the data type, apologies. Strange that it works with SQLite.

No problem ;) I will try it with SQL server, maybe only specific to PostgreSQL

**Comment by deanmarcussen** [comment]
> > No drop of the useless column?
> 
> As i remember you had some problem with some migration steps, and **you commented that it is not safe to remove an index, is it still true?** if so, we can't remove this useless column that is still included in an existing index
> 
> So, is it still a problem to remove an index in a migration step?

You're right. Can't remove it because it is an index. And can't remove the Index because of problems with SQL Server 2012.
Where the `YesSql.DropIndex` command doesn't work on Sql Server 2012, and we caused a bunch of people issues when I did it in (I think) the `AutoroutePartIndex`

I found the old issue https://github.com/OrchardCMS/OrchardCore/issues/6210

Me, I would prefer we say SQL Server 2014 + , but I don't want to cause more issues for people.

So not hard to keep the column

**Comment by Skrypt** [comment]
Ok then steps for this to work is : 

1 - Empty the values in that ScheduledPublishUtc column.
2 - Alter the database colum to use a DateTime.
3 - repopulate them with the JSON data contained in the Document table.

**Comment by jtkech** [comment]
@deanmarcussen  Okay, so i will just rename the new column

I will also try the migration step with SQL server

**Otherwise, if not included in an index, is it safe to remove a column for all databases we support?**

What i didn't tried is to remove a column, and re-add it with the same name but a different type ;)

But i think it would need 2 steps in isolated scopes

**Comment by Skrypt** [comment]
Update my previous comment. Leaving you decide @jtkech

**Comment by Skrypt** [comment]
Well you can always execute SQL directly I guess but it's not the "standard".

**Comment by jtkech** [comment]
So, i asked me the following

> Yes, I asked to me if it is safe to remove and re-create a column of the same name in the same migration scope

And this without removing the index (including this column) because of problems with SQL Server 2012

Here i'm talking about the index created with `CreateIndex()` that is an index of the `PublishLaterPartIndex` table ;)

Just tried, it fails to remove the column saying it is in use

So, nothing more that we can do here

**Comment by Skrypt** [comment]
Well if you empty the table I'm sure it will work. TRUNCATE TABLE PublishLaterPartIndex.

**Comment by Skrypt** [comment]
In this case we could rename PublishLaterPartIndex to PublishLaterPartIndexTemp. Then recreate the PublishLaterPartIndex table. Then move the data from PublishLaterPartIndexTemp to PublishLaterPartIndex. Then DROP PublishLaterPartIndexTemp.

**Comment by jtkech** [comment]
> TRUNCATE TABLE PublishLaterPartIndex.

Yes good idea, but i can't do that with the available methods of yesSql `IAlterTableCommand`

> ... Then DROP PublishLaterPartIndexTemp

With sql 2012 there are problem to use `.DropIndex()`, so not sure we can use `.DropMapIndexTable()`

**Comment by Skrypt** [comment]
Ok if we can't do anything then the issue is that we need to fix these issue first. Also, I'm talking here about using SQL directly in the migration. Can we avoid using anything else than plain SQL to avoid these issues for now?

**Comment by jtkech** [comment]
No problem @Skrypt ;) Yes maybe good to use plain SQL for this kind of things, can you suggest some code doing it, if it is worth doing it for this issue

**Comment by Skrypt** [comment]
The thing is that now we have a requirement to support SQL Server 2012. Which I never knew about. If we go that path then we can't use JSON functions at all. That's one thing.

Also, if in SQL Server 2012 we can't drop a table because of attached indexes then we need to drop the indexes before dropping the table. Now, I can't assert this because I don't have a SQL Server 2012 instance to test with.

Related to this, is it worth supporting every versions of SQL Server? I can't affirm this because normally we should make this work at least with latest versions of each RDB's. So, if someone is still using SQL Server 2012 with Orchard Core in 2021 maybe we should show him SQL Server Express 2019 which is free of use "with some limitations".

**Comment by Skrypt** [comment]
So who is using SQL Server 2012 with Orchard Core?

**Comment by Skrypt** [comment]
So dropping indexes in SQL Server 2012 would be 

``` SQL
DROP INDEX [IDX_AutoroutePartIndex_ContentItemId]"
```

instead of > 2012

``` SQL
DROP INDEX if exists [IDX_AutoroutePartIndex_ContentItemId]"
```

**Comment by jtkech** [comment]
> The thing is that now we have a requirement to support SQL Server 2012. Which I never knew about

Me too ;)

> Also, if in SQL Server 2012 we can't drop a table because of attached indexes then we need to drop the indexes before dropping the table. Now, I can't assert this because I don't have a SQL Server 2012 instance to test with.

There are problems to drop indexes, on my side i have SQL Server 2014

> Related to this, is it worth supporting every versions of SQL Server? I can't affirm this because normally we should make this work at least with latest versions of each RDB's. So, if someone is still using SQL Server 2012 with Orchard Core in 2021 maybe we should show him SQL Server Express 2019 which is free of use "with some limitations".

I agree, if we don't want to support it, it would be worth to know it before finalizing #8245, and also this PR

Hmm, but it seems that there are already a number of users that uses the 2012 version

**Comment by Skrypt** [comment]
``` SQL
-- We create a temp table
CREATE TABLE PublishLaterPartIndexTemp ([Id] int, [DocumentId] [int],
	[ScheduledPublishUtc] [nvarchar](255))

-- We feed the temp table with temp data
INSERT INTO PublishLaterPartIndexTemp (DocumentId, ScheduledPublisUtc)
SELECT DocumentId, ScheduledPublishUtc FROM PublishLaterPartIndex

DROP INDEX IDX_PublishLaterPartIndex_ScheduledPublishUtc ON PublishLaterPartIndex
ALTER PublishLaterPartIndex
DROP COLUMN ScheduledPublishUtc;

ALTER PublishLaterPartIndex
ADD COLUMN ScheduledPublishUtc DateTime;
CREATE NONCLUSTERED INDEX IDX_PublishLaterPartIndex_ScheduledPublishUtc ON PublishLaterPartIndex (ScheduledPublishUtc ASC)

-- We move back the data from the temp table
INSERT INTO PublishLaterPartIndex (DocumentId, ScheduledPublisUtc)
SELECT DocumentId, CONVERT(DATETIME, ScheduledPublishUtc) FROM PublishLaterPartIndexTemp

-- We drop the temp table
DROP TABLE PublishLaterPartIndexTemp
```

Now with this approach you only need to be able to

**Comment by Skrypt** [comment]
By the way JSON Functions are SQL Server 2016 and later.

https://docs.microsoft.com/en-us/sql/t-sql/functions/json-functions-transact-sql?view=sql-server-ver15

**Comment by Skrypt** [comment]
Ok, I've been thinking about SQL Script migrations this afternoon and I came up with the evidence that SQLite can't run complex syntax scripts. We can't for example do a WHILE or FOR in SQLite which restrict us at only using YesSQL for doing DB migrations.

Now, though, each of the DB providers we have allows for using JSON FUNCTIONS to retrieve data from their tables. So, there is no issue about retrieving JSON data. The issue is just about when we want to batch execute queries that INSERT/UPDATE data.

So, all the DB providers allows us to use WHILE and FOR statements except SQLite.
So, composing a SQL migration for each of these RDB's could be one option but one that is time spending. Right now, the only thing I can see is that we need to maybe think about a specific method to be able to do complex migrations with SQLite apart than all the others which can be done with SQL Scripting.

**Comment by jtkech** [comment]
@Skrypt Yes, when some alterations are not supported by all providers, and / or if we need to swap data, would be good to have some dedicated scripts that would run acoording to the current provider. But looks quite complex to do, but maybe worth to start working on it, so that we could use it afterwards. E.g. if on a new installation we created optimized indexes, but incomplete ones in the related migration because of column sizes, we could suggest afterwards a migration to upgrade it, meanwhile people could also do some of the work manually on their own database.

**Pull request merged**